### PR TITLE
optional!

### PR DESCRIPTION
**Pull request by jmervine**
Adding `optional!` for required if requested type functionality.

### CODE REVIEW

**Comment by reidmix** [comment]
This is what required was supposed to do... looks like it got lost on extraction:
https://github.com/interagent/pliny/blob/master/lib/pliny/config_helpers.rb#L85

Pliny always seemed a little over-coded, what I'd do is:
- get rid of `optional!`, `create!`
- have create take on an option to raise if required, (ie. `required: true` and default it to false)
- have required use that option 

I think that gets you what you want.

**Comment by jmervine** [comment]
I like the idea of `required` raising on startup, preventing an application from starting with missing configuration is a valid and important use-case IMO.

Perhaps we leave `optional` as is, and adapt `required` to support an `on_request` param, which raises when it's requested instead of when loaded.

Or flip it, have an `on_load` param, which raises on load if it's set to `true`?

Also, to be backwards compatible w/ Pliny, I'd like to alias `required` to to `mandatory`, unless there are any objections.

**Comment by jmervine** [comment]
Wrapping up my thoughts on this...
- `required` raise on request
- `required!` raise on load
- `required :key, on_load: true` raise on load
- `manditory` aliased to `required` for backwards compatability
- `optional` as it
- `override` as is

@reidmix @joshuatobin thoughts?

**Comment by reidmix** [comment]
Too complicated IMO.  

Just implement `required` to raise on startup.  Raising on first request sounds like a deferment to the worst time.  No need for bang methods then.

Get rid of manditory, since we're creating something new, lets not bog it down with other apps old crap <3 just update the old crap to use the new terminology :D

**Comment by jmervine** [comment]
Well, it's currently raising on startup as is. So that would be a `noop`. And deferring to first request is exactly what I need for another project. :P

**Comment by reidmix** [comment]
Ah I understand now.  Before I saw that the raise is not actually implemented for `required` which is a mistake (see my first comment)... so I thought that's why you were implementing `optional!`. 

So yeah, I guess I'm a -1 on this PR, sorry. :P  But I do think that `required` should do what you are wanting before it becomes a problem during runtime execution.  

Please do fix `required`.

**Comment by jmervine** [comment]
@reidmix Refactored as discussed. cc @joshuatobin

File: README.md [hunk 1]
```diff
@@ -15,6 +17,17 @@ $ gem install configerator
...
+### Methods
+
+* `required :key`
+    * Require a key, raise a `KeyError` if key is not supplied on application start up.
+* `requied :key, error_on_load: false`
+    * Require a key, raise a `RuntimeError` if key is not supplied when `key` is requested.
+* `optional :key`
+    * Create `key`, set to `nil` if not present.
+* `override :key, :value`
+    * Create `key`, set to `value` if not present.
```
Comment (joshuatobin) [inline README.md:29]: +1 these all seem to make sense.

**Comment by joshuatobin** [comment]
👍

**Comment by reidmix** [comment]
👍 lgtm

**Pull request merged**